File: Java/Dynamic-Programming/TravellingSalesMan.java
import java.util.Scanner;

public class TravellingSalesMan {
    static int findHamiltonianCycle(int[][] distance, boolean[] visitCity, int currPos, int cities, int count, int cost, int hamiltonianCycle) {
        if (count == cities && distance[currPos][0] > 0) {
            hamiltonianCycle = Math.min(hamiltonianCycle, cost + distance[currPos][0]);
            return hamiltonianCycle;
        }

        for (int i = 0; i < cities; i++) {
            if (visitCity[i] == false && distance[currPos][i] > 0) {
                visitCity[i] = true;
                hamiltonianCycle = findHamiltonianCycle(distance, visitCity, i, cities, count + 1, cost + distance[currPos][i], hamiltonianCycle);
                visitCity[i] = false;
            }
        }
        return hamiltonianCycle;
    }

    public static void main(String[] args) {
        int cities;
        Scanner sc = new Scanner(System.in);
        System.out.println("Enter total number of cities: ");
        cities = sc.nextInt();

        int[][] distance = new int[cities][cities];
        for (int i = 0; i < cities; i++) {
            for (int j = 0; j < cities; j++) {
                System.out.println("Distance from city " + (i + 1) + " to city " + (j + 1) + ": ");
                distance[i][j] = sc.nextInt();
            }
        }

        boolean[] visitCity = new boolean[cities];
        visitCity[0] = true;

        int hamiltonianCycle = Integer.MAX_VALUE;
        hamiltonianCycle = findHamiltonianCycle(distance, visitCity, 0, cities, 1, 0, hamiltonianCycle);

        System.out.println("The minimum weighted Hamiltonian Cycle is: " + hamiltonianCycle);

        System.out.print("Optimal path: 1");
        int nextCity = 0;
        for (int i = 1; i < cities; i++) {
            for (int j = 0; j < cities; j++) {
                if (distance[nextCity][j] > 0 && distance[nextCity][j] + distance[j][0] == hamiltonianCycle) {
                    System.out.print(" -> " + (j + 1));
                    nextCity = j;
                    break;
                }
            }
        }
        System.out.println(" -> 1");
        sc.close();
    }
}

File: Python/Dynamic-Programming/TravellingSalesMan.py
import itertools

def calculate_distance(city1, city2):
    # Calculate the Euclidean distance between two cities
    return ((city1[0] - city2[0])**2 + (city1[1] - city2[1])**2) ** 0.5

def calculate_total_distance(path, cities):
    # Calculate the total distance for a given path
    total_distance = 0
    for i in range(len(path)-1):
        city1 = cities[path[i]]
        city2 = cities[path[i+1]]
        total_distance += calculate_distance(city1, city2)
    return total_distance

def find_minimum_distance(cities):
    # Generate all possible permutations of cities
    all_paths = list(itertools.permutations(range(len(cities))))

    # Initialize variables to track the minimum distance and path
    min_distance = float('inf')
    min_path = []

    # Iterate through all paths and find the minimum distance
    for path in all_paths:
        distance = calculate_total_distance(path, cities)
        if distance < min_distance:
            min_distance = distance
            min_path = path

    return min_path, min_distance

def main():
    # Get the total number of cities from the user
    num_cities = int(input("Enter the total number of cities: "))

    # Initialize an empty list to store the cities' coordinates
    cities = []

    # Get the coordinates of each city from the user
    for i in range(num_cities):
        x = float(input(f"Enter the x-coordinate of city {i+1}: "))
        y = float(input(f"Enter the y-coordinate of city {i+1}: "))
        cities.append((x, y))

    # Find the minimum distance and path
    min_path, min_distance = find_minimum_distance(cities)

    # Print the minimum distance and path
    print("The minimum distance is:", min_distance)
    print("The optimal path is:", min_path)

if __name__ == "__main__":
    main()


